lib/src/uri_template.dart

```diff
@@ -148,7 +151,7 @@ class UriParser extends UriPattern {
       for (var key in _queryVariables.keys) {
         var value = _queryVariables[key];
         if (value == null) {
-          if (!uri.queryParameters.containsKey(key)) {
+          if (!_queryParamsAreOptional && !uri.queryParameters.containsKey(key)) {
```

Comment:
this might be clearer if you swap the if cases:
```
if (_queryParamsAreOptional || uri.queryParameters.containsKey(key)) {
  parameters[key] = uri.queryParameters[key];
} else {
  return null;
}
``